R/stat-density.r

```diff
@@ -125,6 +127,17 @@ compute_density <- function(x, w, from, to, bw = "nrd0", adjust = 1,
   dens <- stats::density(x, weights = w, bw = bw, adjust = adjust,
     kernel = kernel, n = n, from = from, to = to)
 
+  integrate_density <- function(x, y) {
+    x_diff <- diff(x)
+    x_diff <- c(x_diff[1], x_diff)
+    sum(y * x_diff, na.rm = TRUE)
+  }
+
+  total_pre <- integrate_density(dens$x, dens$y)
+  total_post <- integrate_density(trans$inverse(dens$x), dens$y)
+
+  dens$y <- dens$y * total_pre / total_post
+
```

Comment:
A few observations

1. If `stats::density` guarantees that the area is functionally 1, then `total_pre` is not needed.
2. Instead of `integrate_density` why not hide all this diligence under a `rescale_density` function.
3. Rescale the density only if the `trans` is not an identity transformation.

Reply:
Thanks, Hassan. For now, I'm just hoping to understand if the ggplot authors think we should be rescaling the density. I don't expect this will be the final implementation.

Having said that, you've raised some excellent points. I'll address them below:

> If `stats::density` guarantees that the area is functionally 1, then `total_pre` is not needed.

Unfortunately, `stats::density()` doesn't guarantee this. In fact, the output is a linear interpolation of a fast Fourier transform, so it's very unlikely the area equals 1. Given this, we need to take a ratio of the areas before and after the transformation.

> Instead of `integrate_density` why not hide all this diligence under a `rescale_density` function.

When it comes to integrating the area under the curve, there are quite a few choices to make (e.g., use `stats::integrate()`, use the [trapezoidal rule](https://en.wikipedia.org/wiki/Trapezoidal_rule), etc). I've just done a very simple approximation (the [Riemann sum](https://en.wikipedia.org/wiki/Riemann_sum)), and I fully expect this will be replaced. So my 1integrate_density()` function is acting as a placeholder of sorts.

> Rescale the density only if the trans is not an identity transformation.

I totally agree on this. We shouldn't waste CPU time when there's no transformation to be performed. I'm not sure of the best way to identify this though, so I'd love feedback on how to achieve this.